File: plugins/git-utils/src/core/jira.ts
// ============================================================
// JiraService — Jira 티켓 감지
// ============================================================
// detect-jira-ticket.sh 의 3단계 regex 매칭 로직을 통합합니다.
//
// 지원 패턴:
//   - feat/WAD-0212         → WAD-0212
//   - feat/wad-0212         → WAD-0212 (대문자 정규화)
//   - WAD-0212              → WAD-0212
//   - fix/wad-2223/desc     → WAD-2223
//
// 기존 대비 개선: 프로젝트 키 최소 2글자 제한으로 오매칭 방지
// ============================================================

import type { JiraTicket } from '../types';

export interface JiraService {
  detectTicket(branchName: string): JiraTicket | null;
}

// Pattern 1: prefix/TICKET-123 또는 prefix-TICKET-123
const PREFIXED_TICKET = /^[a-z]+[-/]([A-Za-z]{2,}-\d+)/;

// Pattern 2: 대문자 TICKET-123 (브랜치명 어디서든)
const UPPERCASE_TICKET = /([A-Z]{2,}-\d+)/;

// Pattern 3: 소문자 ticket-123 (브랜치명 어디서든, 최소 2글자 프로젝트 키)
const LOWERCASE_TICKET = /([a-z]{2,}-\d+)/;

export function createJiraService(): JiraService {
  return {
    detectTicket(branchName: string): JiraTicket | null {
      if (!branchName) return null;

      // Pattern 1: prefix/TICKET-123
      let match = branchName.match(PREFIXED_TICKET);
      if (match) {
        const raw = match[1];
        return { raw, normalized: raw.toUpperCase() };
      }

      // Pattern 2: UPPERCASE TICKET-123
      match = branchName.match(UPPERCASE_TICKET);
      if (match) {
        return { raw: match[1], normalized: match[1] };
      }

      // Pattern 3: lowercase ticket-123
      match = branchName.match(LOWERCASE_TICKET);
      if (match) {
        const raw = match[1];
        return { raw, normalized: raw.toUpperCase() };
      }

      return null;
    },
  };
}

File: plugins/atelier/cli/src/git/core/jira.rs
//! Jira ticket detection, ported from git-utils `core/jira.ts`
//! (originally the 3-stage regex matching of `detect-jira-ticket.sh`).
//!
//! Supported patterns:
//!   - `feat/WAD-0212`     → `WAD-0212`
//!   - `feat/wad-0212`     → `WAD-0212` (uppercased)
//!   - `WAD-0212`          → `WAD-0212`
//!   - `fix/wad-2223/desc` → `WAD-2223`
//!
//! The project key requires at least two letters to avoid mis-matches.

use regex::Regex;
use std::sync::LazyLock;

/// A detected Jira ticket: the `raw` substring that matched and its
/// `normalized` (uppercased) form.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct JiraTicket {
    pub raw: String,
    pub normalized: String,
}

// Pattern 1: prefix/TICKET-123 or prefix-TICKET-123 (lowercase prefix).
static PREFIXED_TICKET: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^[a-z]+[-/]([A-Za-z]{2,}-\d+)").unwrap());
// Pattern 2: UPPERCASE TICKET-123 anywhere in the branch name.
static UPPERCASE_TICKET: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"([A-Z]{2,}-\d+)").unwrap());
// Pattern 3: lowercase ticket-123 anywhere (≥2-letter project key).
static LOWERCASE_TICKET: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"([a-z]{2,}-\d+)").unwrap());

/// Detects a Jira ticket in `branch_name`, returning `None` when no pattern
/// matches. Mirrors `JiraService.detectTicket` from git-utils.
pub fn detect_ticket(branch_name: &str) -> Option<JiraTicket> {
    if branch_name.is_empty() {
        return None;
    }

    // Pattern 1: prefix/TICKET-123 → normalize to uppercase.
    if let Some(caps) = PREFIXED_TICKET.captures(branch_name) {
        let raw = caps[1].to_string();
        let normalized = raw.to_uppercase();
        return Some(JiraTicket { raw, normalized });
    }

    // Pattern 2: already-uppercase TICKET-123 → raw == normalized.
    if let Some(caps) = UPPERCASE_TICKET.captures(branch_name) {
        let raw = caps[1].to_string();
        return Some(JiraTicket {
            normalized: raw.clone(),
            raw,
        });
    }

    // Pattern 3: lowercase ticket-123 → normalize to uppercase.
    if let Some(caps) = LOWERCASE_TICKET.captures(branch_name) {
        let raw = caps[1].to_string();
        let normalized = raw.to_uppercase();
        return Some(JiraTicket { raw, normalized });
    }

    None
}
